Ios migration guide (#45)
MapsPeople/docsite: docs/01-getting-started/ios/v4/display-a-map.mdx

---
title: Display a Map
hide_title: false
hide_table_of_contents: true
sidebar_position: 3
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Now that we have the prerequisite API keys, and the project set up, we can start adding basic functionality to the app. We will start by having the app display a map.

## Display a Map with MapsIndoors

:::tip async

Please note that data in MapsIndoors is loaded asynchronously. This results in behavior where data might not have finished loading if you call methods accessing it immediately after initializing MapsIndoors. Best practice is to set up `listeners` or `delegates` to inform of when data is ready. Please be aware of this when developing using the MapsIndoors SDK.

:::

> 

Reviewer: This should be alleviated in v4, so no need to mention here any longer.

> In order to accomplish this we will be utilising the [MPMapControl Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/interface_m_p_map_control.html). Open the file, `ViewController.swift` and, once again, add in the following import statements to the top of the file,

Reviewer: ```suggestion
In order to accomplish this we will be utilising the [MPMapControl Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v4-doc/documentation/mapsindoors/mpmapcontrol). Open the file, `ViewController.swift` and, once again, add in the following import statements to the top of the file,
```

### Import MapsIndoors

<Tabs groupId="getting-started-ios-v4-map">
<TabItem value="swift-google" label="Swift - Google Maps" default>

```swift
import GoogleMaps  

> import MapsIndoors

Reviewer: ```suggestion
import MapsIndoorsCore
```

```

</TabItem>
<TabItem value="swift-mapbox" label="Swift - Mapbox" default>

```swift
import Mapbox  

> import MapsIndoors

Reviewer: ```suggestion
import MapsIndoorsCore
```

                if let firstLocation = locations.first {
                    mapControl.select(location: firstLocation, behavior: .default)
                    mapControl.select(floorIndex: firstLocation.floorIndex.intValue)
                }
            }
        }
```

</TabItem>
<TabItem value="swift-mapbox" label="Swift - Mapbox" default>

```swift
        // Set camera then create MapControl
        Task {
            // pass the previously created MapConfig to `createMapControl`
            if let mapControl = MPMapsIndoors.createMapControl(mapConfig: mapConfig) {
                
                let query = MPQuery()
                let filter = MPFilter()
                
                query.query = "White House"
                filter.take = 1
                
                let locations = await MPMapsIndoors.shared.locationsWith(query: query, filter: filter)
                if let firstLocation = locations.first {
                    mapControl.select(location: firstLocation, behavior: .default)
                    mapControl.select(floorIndex: firstLocation.floorIndex.intValue)
                }
            }
        }
```

</TabItem>
</Tabs>

> The [MPMapControl Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/interface_m_p_map_control.html) the is connective class between the Map Engine(either Google Maps or Mapbox) and the MapsIndoors. It allows the two services to collaborate by overlaying the MapsIndoors information over the Map Engine information.

Reviewer: ```suggestion
The [MPMapControl Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v4-doc/documentation/mapsindoors/mpmapcontrol) is the connective class between the Map Engine (either Google Maps or Mapbox) and MapsIndoors. It allows the two services to collaborate by overlaying the MapsIndoors information on the Map Engine information.
```

> *Note: This also means logic is appended onto many Google Maps listeners, which means that using Google Maps listeners directly might break intended behavior of the MapsIndoors experience. We recommend checking our [reference docs](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/index.html), and see if you can add a specific `Listener` through the `MapControl`. If this is possible, it is highly recommend you do so.*

Reviewer: Must also mention Mapbox

> *Note: This also means logic is appended onto many Google Maps listeners, which means that using Google Maps listeners directly might break intended behavior of the MapsIndoors experience. We recommend checking our [reference docs](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/index.html), and see if you can add a specific `Listener` through the `MapControl`. If this is possible, it is highly recommend you do so.*

Reviewer: ```suggestion
*Note: This also means logic is appended onto many Google Maps listeners, which means that using Google Maps listeners directly might break intended behavior of the MapsIndoors experience. We recommend checking our [reference docs](https://app.mapsindoors.com/mapsindoors/reference/ios/v4-doc/documentation/mapsindoors), and see if you can add a specific `Listener` through `MPMapControl`. If this is possible, it is highly recommend you do so.*
```

What this means is that we should first create our Map, that is why in the previous `Set up Your Environment` we did so for `Mapbox`. This is how to do for `GoogleMaps`:

```swift
// inside `viewDidLoad()`
self.view = mapView
// where mapView is of type `GMSMapView`
 ````

> Running the app wihtout the use of code aobve indeed displays a map, however this is through the use of the Map Engine of your choice exclusively and with a default map region displayed. But we used MapsIndoors to showcase a specific location.

Reviewer: ```suggestion
Running the app without the use of code above indeed displays a map, however this is only using the Map Engine of your choice and with a default map region displayed. But we want to use MapsIndoors to showcase a specific location.
```

> We have now added a *very* simple search feature! Running the app now should yield a combined map of The White House, showing both the external and internal geographical information. However, let us try and understand what is actually happening.

Reviewer: This seems kind of out of place…

We initialise the mapControl for MapsIndoors with respect to the `mapView` generated by the Google Maps API.

> Afterwards, we set-up a Query and Filter based on the [MPQuery Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/interface_m_p_query.html#a9da8ff62b6d17e45403b9005c73e811f) and [MPFilter Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/interface_m_p_filter.html) classes respectively, these serve as our search specifications.

Reviewer: ```suggestion
Afterwards, we set up a Query and Filter based on the [MPQuery Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v4-doc/documentation/mapsindoors/mpquery) and [MPFilter Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v4-doc/documentation/mapsindoors/mpfilter) classes respectively, these serve as our search specifications.
```

> Finally, the [MPLocationService Class](https://app.mapsindoors.com/mapsindoors/reference/ios/v3/interface_m_p_location_service.html#a140844121a239dcf1709210e1723c312) enables us to search through the full collection of Locations based on our previously established query and filter, such that we can go to that location on the map. Moving the location is all handled by mapControl as it handles it for both Google Maps and MapIndoors.

Reviewer: MPLocationService is no longer used directly. Instead, MPMapsIndoors.shared.locationsWith(query:filter:) is used.
Please rephrase paragraph.

Link to ref docs: https://app.mapsindoors.com/mapsindoors/reference/ios/v4-doc/documentation/mapsindoors/mapsindoorsshared/locationswith(query:filter:)

## Expected Result

<img src="/img/getting-started/ios_display-a-map.png" alt="An animation showing the desired behaviour of this tutorial" width="350"/>

Feel free to change the query from "White House" to a known building in *your* MapsIndoors dataset.

We have now added a search feature to the app, albeit a search feature which the user cannot interact with. Next, let us look into how we can let the user interact with the map through a search bar.
### Complete code

<Tabs groupId="getting-started-ios-v4-map">
<TabItem value="swift-google" label="Swift - Google Maps" default>

```swift
import GoogleMaps  

> import MapsIndoors

Reviewer: ```suggestion
import MapsIndoorsCore
import MapsIndoorsGoogleMaps
```

// In your `AppDelegate`
@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    static let mApiKey = [YOUR_MAPSINDOORS_API_KEY]
    static let gApiKey = ProcessInfo.processInfo.environment[YOUR_GOOGLE_API_KEY]

    var window: UIWindow?
    

>     var currentAuthorizationFlow: OIDExternalUserAgentSession?

Reviewer: Too much copied directly from our Demos app 😄 
```suggestion
```

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

>         GMSServices.provideAPIKey(AppDelegate.gApiKey)

Reviewer: We have not mentioned `AppDelegate.gApiKey` so we should not make use of it here (even if we do in our Demos app)

        Task {
            await MPMapsIndoors.shared.load(apiKey: AppDelegate.mApiKey)
        }
        return true
    }
    
    // ...
}

// In your `ViewController`

class ViewController: UIViewController, GMSMapViewDelegate {
    var mapView: GMSMapView? = nil
    var mapControl: MPMapControl? = nil
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let camera = GMSCameraPosition.camera(withLatitude: 57.057964, longitude: 9.9504112, zoom: 20)
        mapView = GMSMapView.map(withFrame: .zero, camera: camera)
        self.view = mapView

>         let location = CLLocationCoordinate2D(latitude: 57.057964, longitude: 9.9504112)

Reviewer: The `location` var is never used
```suggestion
```

        
        Task {
            let error = await MPMapsIndoors.shared.load(apiKey: [YOUR_MAPSINDOORS_API_KEY])
            
            if let mapControl = MPMapsIndoors.createMapControl(mapView: mapView!) {
                let query = MPQuery()
                let filter = MPFilter()
                
                query.query = "White House"
                filter.take = 1
                
                let locations = await MPMapsIndoors.shared.locationsWith(query: query, filter: filter)
                if let firstLocation = locations.first {
                    mapControl.select(location: firstLocation, behavior: .default)
                    mapControl.select(floorIndex: firstLocation.floorIndex.intValue)
                }
            }
        }
    }
}

```

</TabItem>
<TabItem value="swift-mapbox" label="Swift - Mapbox" default>

```swift
import Mapbox  

> import MapsIndoors

Reviewer: ```suggestion
import MapsIndoorsCore
import MapsIndoorsMapbox
```
